pkgs/by-name/ha/hare/setup-hook.sh

```diff
@@ -1,9 +1,26 @@
-addHarepath () {
+readonly hareDefaultFlags=(@hare_default_flags@)
+
+setHareEnv() {
+    HARECACHE="$(mktemp -d)"
+    export HARECACHE
+    echoCmd 'HARECACHE' "$HARECACHE"
+    HAREFLAGS="${HAREFLAGS-} ${hareDefaultFlags[*]}"
+    export HAREFLAGS
+    echoCmd 'HAREFLAGS' "$HAREFLAGS"
+    makeFlagsArray+=(
+        HARECACHE="$HARECACHE"
+        HAREFLAGS="$HAREFLAGS"
+    )
+    echoCmd 'makeFlagsArray' "${makeFlagsArray[@]}"
```

Comment:
it's unfortunate that plumbing `HAREFLAGS` through to the actual `hare` invocation requires cooperation from every layer (correct me if i'm wrong): not only do we have to call `make HAREFLAGS="$HAREFLAGS"`, but also the project's Makefile needs to be aware of that and invoke hare with `hare $(HAREFLAGS)`. then as more hare projects come along using things other than make (like haredo), we have to make sure each new build tool is properly integrated...

i wonder if we should actually just wrap the `hare` executable so that it appends `$HAREFLAGS` (or even our own `NIX_HAREFLAGS`) directly, and then just let each package/builder do its own thing rather than try to plumb things through each layer cooperatively?

Reply:
I like your idea of wrapping the hare executable: We avoid the need for integration with each new build tool, while achieving the same, that is, passing HAREFLAGS to the executable.

I also think that we can split what's being done in the setup-hook:

1. `HARECACHE` being only an environment variable still allows the hare binary to use, while also allowing build tools to have access to its value;
2. The *default* `HAREFLAGS` now are `-q`, for not polluting the build logs; `-R`, for building on release mode; and `-a<arch>`, for cross-compiling. I believe we should add a `hareDontUseReleaseMode` --- or something equivalent --- attribute, defaulting to `false`, to allow derivation to disable the `-R`, since, although rare, it may rely on the `debug` module.

Reply:
I've tried a few approaches regarding wrapping the hare binary to no avail:

1. Creating a `hare() {}` bash function: It works on packages the call directly from the shell environment --- *e. g.*, [haredo](https://github.com/NixOS/nixpkgs/blob/85d3f97277c0a0b5b542951e8990b4bd2bbe2a26/pkgs/by-name/ha/haredoc/package.nix#L30) and [haredoc](https://github.com/NixOS/nixpkgs/blob/85d3f97277c0a0b5b542951e8990b4bd2bbe2a26/pkgs/by-name/ha/haredoc/package.nix#L30) ---, but not for packages that call it from a build tool --- *e. g.*, [bonsai](https://github.com/NixOS/nixpkgs/blob/85d3f97277c0a0b5b542951e8990b4bd2bbe2a26/pkgs/by-name/bo/bonsai/package.nix) with `make` ---, so it's a no go.
2. Creating `writeShellApplication` with the hare binary wrapped in it and adding it to `PATH` through the setup-hook the same way `HAREPATH` is set: Although the script ends up in `PATH`, it ends up after hare's derivation path, thus the latter has precedence over the former, which ends up not being used.

Any ideas on this, @uninsane? Perhaps we could try again a `hare.hook` attribute passed through `passthru`, since it would allow us to not have the unwrapped hare binary `PATH`, but we would still need to solve the cross-compilation brought upon the first iteration.

Reply:
my first thought would be to use `makeWrapper` when building hare itself:
```nix
# hare/package.nix:
{ ... }:
stdenv.mkDerivation {
  pname = "hare";
  ...
  nativeBuildInputs = [ makeWrapper ... ];
  postInstall = ''
    wrapProgram $out/bin/hare \
      --add-flags '$NIX_HAREFLAGS'  # quoted so it won't expand at the time we make the program, but hopefully at the time the wrapper is invoked
  '';
};
```

like, don't even create a new derivation for this: just add that `postInstall` to the existing `hare` derivation. if that doesn't fit the bill i would have to look at our `cc` wrapper for inspiration.

Reply:
Neither `--add-flags` nor `--append-flags` would work because the added `$NIX_HAREFLAGS` would be, respectively, before the subcommand --- *e. g.*, `hare $NIX_HAREFLAGS <cmd>` ---, causing a `unrecognized option: ` error, or after all arguments --- *e. g.*, `hare <cmd> $@ $NIX_HAREFLAGS` ---, which would make it be treated as an positional argument any time a path was passed and we cannot guarantee that every package invocation of the hare binary would be called without passing a path to it --- like so: `hare <cmd> <flags>`.

If there is a way to select at which index the appended flag would go, we would be able to use your `wrapProgram` solution.

Reply:
I've tried creating a different derivation and adding it to `pkgs/top-level/all-packages.nix` as `hareHook` and it works with both native and cross-compilation; it, however, when set as an attribute of hare's `passthru` has the same issue with cross-compilation that the `hare.hook` solution had. Moreover, if we were to change to a different derivation, the current `setupHook` of hare would had to go to it and it would break hare's functionality on a `nix-shell`, since:

1. For `HAREPATH` to be set, the user would had to add `hareHook` to the nix-shell;
2. We would also need to find a way to disabling passing `-qRa<arch>` --- perhaps a pattern like `noDefaultFlags` ---, since, in a nix-shell, the user would prefer fine control of which options are passed to the binary.

Reply:
as for the `noDefaultFlags` stuff, i notice several build hooks support attributes like `mesonBuildType = "plain"|"debug"|"release"|...` (e.g. `cmakeBuildType` for cmake, `cargoBuildType` for rust, `dubBuildType` for Dlang),  so `hareBuildType = "plain"|"release"` may be an idiomatic way to control whether we pass the `-R` flag or not.

hopefully i'll have some time to dig into the wrapper stuff this weekend.